core/src/main/java/org/elasticsearch/search/highlight/HighlightPhase.java

```diff
@@ -102,6 +105,20 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
                     continue;
                 }
 
+                // We should prevent highlighting if a field is anything but a text or keyword field.
+                // However, someone might implement a custom field type that has text and still want to
+                // highlight on that. We cannot know in advance if the highlighter will be able to
+                // highlight such a field and so we do the following:
+                // If the field is only highlighted because the field matches a wildcard we assume
+                // it was a mistake and do not process it.
+                // If the field was explicitly given we assume that whoever issued the query knew
+                // what they were doing and try to highlight anyway.
+                if (fieldNameContainsWildcards) {
+                    if (fieldMapper.fieldType().typeName().equals(TextFieldMapper.CONTENT_TYPE) == false && fieldMapper.fieldType().typeName().equals
+                        (KeywordFieldMapper.CONTENT_TYPE) == false) {
```

Comment:
like Clint said, we should allow string fields too (which will continue to exist on 2.x indices)